### PR TITLE
[analyzer] Unbreak [[clang::suppress]] on checkers without decl-with-issue.

### PR DESCRIPTION
**Pull request by haoNoQ**
There are currently a few checkers that don't fill in the bug report's "decl-with-issue" field (typically a function in which the bug is found).

The new attribute `[[clang::suppress]]` uses decl-with-issue to reduce the size of the suppression source range map so that it didn't need to do that for the entire translation unit.

I'm already seeing a few problems with this approach so I'll probably redesign it in some point as it looks like a premature optimization. Not only checkers shouldn't be required to pass decl-with-issue (consider clang-tidy checkers that never had such notion), but also it's not necessarily uniquely determined (consider leak suppressions at allocation site).

For now I'm adding a simple stop-gap solution that falls back to building the suppression map for the entire TU whenever decl-with-issue isn't specified. Which won't happen in the default setup because luckily all default checkers do provide decl-with-issue.

### CODE REVIEW

**Comment by haoNoQ** [comment]
The affected checkers were:
- Checkers in the `webkit` package (hence cc @jkorous-apple). These checkers manually scan the entire TU in order (rejecting the comforts of `checkASTCodeBody`) in order to properly cover uninstantiated templates. There's no reason they can't provide a decl-with-issue but it's often annoying to detect.
- Copy-paste error checkers. Again, the checker can provide decl-with-issue, but plumbing gets quite annoying. The test case is super funny though!

**Review by Xazax-hun — APPROVED** [review]

**Review by steakhal — COMMENTED** [review]
I only have minor nits. No objections.

**Review by jkorous-apple — APPROVED** [review]
Thanks for fixing this!
LGTM from the perspective of WebKit checkers.

File: clang/include/clang/StaticAnalyzer/Core/BugReporter/BugSuppression.h [hunk 1]
```diff
@@ -27,6 +28,8 @@ class PathDiagnosticLocation;
 
 class BugSuppression {
 public:
+  BugSuppression(ASTContext &ACtx) : ACtx(ACtx) {}
```
[the remark this answers is not in the document]
Comment (haoNoQ) [reply clang/include/clang/StaticAnalyzer/Core/BugReporter/BugSuppression.h:31]: Hmm I somehow never noticed that we have any actual const-correctness for `ASTContext`, thanks!

**Review by steakhal — APPROVED** [review]

**Pull request merged**